File: src/main/scala/org/rasterfun/util/RasterPanel.scala
package org.rasterfun.util

import javax.swing.JPanel
import java.awt.Graphics
import java.awt.event.{ComponentAdapter, ComponentEvent}
import org.rasterfun.{Area}
import org.rasterfun.functions.{Fun}

/**
 * 
 */
class RasterPanel(initialArea: Area, red: Fun, green: Fun, blue: Fun, alpha: Fun) extends JPanel {

  private var fastImage: FastImage = null
  private var imagePainted = false
  private var area: Area = initialArea

  addComponentListener(new ComponentAdapter {
    override def componentResized(e: ComponentEvent) = {
      fastImage = new FastImage(getWidth, getHeight)
      imagePainted = false
    }
  })


  def calculateBuffer(buffer: Array[Int], width: Int, height: Int, area: Area,
                      red: Fun, green: Fun, blue: Fun, alpha: Fun) {
    require(buffer.length == width * height, "Buffer size should match width * heigth")

    def clampScale(v: Float): Int = {
      if (v < 0) 0
      else if (v > 1) 0xff
      else (v * 0xff).toInt & 0xff
    }

    val sxStep = area.width / width
    val syStep = area.height / height
    val sampleSize = (sxStep + syStep) * 0.5f

    var index = 0
    var sy = area.minY
    var y = 0
    while (y < height) {

      var sx = area.minX
      var x = 0
      while (x < width) {
        val r = clampScale(red(sx, sy, sampleSize))
        val g = clampScale(green(sx, sy, sampleSize))
        val b = clampScale(blue(sx, sy, sampleSize))
        val a = clampScale(alpha(sx, sy, sampleSize))
        buffer(index) =
                (a << 24) |
                (r << 16) |
                (g << 8) |
                (b << 0)

        sx += sxStep
        x += 1
        index += 1
      }

      sy += syStep
      y += 1
    }
  }


  override def paintComponent(g: Graphics): Unit = {
    if (!imagePainted) {

      calculateBuffer(fastImage.buffer, fastImage.width, fastImage.height, area, red, green, blue, alpha)
      imagePainted = true
    }


    fastImage.renderToGraphics(g)
  }

}

File: src/main/java/org/rasterfun/utils/RasterPanel.java
package org.rasterfun.utils;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;

/**
 * A Swing Panel that shows a backing FastImage.
 */
public class RasterPanel extends JPanel {

    private FastImageRenderer renderer;
    private FastImage fastImage = null;
    private boolean imagePainted = false;

    public RasterPanel() {
        this(null);

        setOpaque(true);
    }

    /**
     * @param renderer something that can render the contents of the image when needed.
     */
    public RasterPanel(FastImageRenderer renderer) {
        this.renderer = renderer;

        // Listen to resizes, and re-render the image when a resize happened
        addComponentListener(new ComponentAdapter() {
            public void componentResized(ComponentEvent e) {

                final int width = getWidth();
                final int height = getHeight();

                // Create image if we have any size to work with.
                if (width > 0 && height > 0) fastImage = new FastImage(width, height);
                else fastImage = null;

                // Notify renderer about the new size
                if (RasterPanel.this.renderer != null) {
                    RasterPanel.this.renderer.setViewSize(width, height);
                }

                reRender();
            }
        });
    }

    /**
     * @return the renderer used to render this image.
     */
    public FastImageRenderer getRenderer() {
        return renderer;
    }

    /**
     * @param renderer the renderer to use for rendering this image.
     */
    public void setRenderer(FastImageRenderer renderer) {
        this.renderer = renderer;
        renderer.setViewSize(getWidth(), getHeight());
        reRender();
    }

    /**
     * Triggers a re-render of the image and a repaint of this panel.
     */
    public void reRender() {
        imagePainted = false;
        repaint();
    }

    public void paintComponent(Graphics g) {
        if (fastImage != null) {

            // Re-render the image if needed
            if (!imagePainted && renderer != null) {
                renderer.renderImage(fastImage);
                imagePainted = true;

                // Flush the image to be sure we have the latest version
                fastImage.getImage().flush();
            }

            // Draw the renderer image to the panel
            fastImage.renderToGraphics(g);
        }
    }

}